### PR TITLE
Applayer plugin 5053 v3.3

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
Preliminary work for https://redmine.openinfosecfoundation.org/issues/5053

Describe changes:
- get ready to use dynamic number of app-layer protos (also work with static constant) in some places
- preventive fix of macro with parenthesis cc @jufajardini 

#11572 next round

#11696 with clean git history and green CI

Still more work to do after : 
Only `AppProtoStrings` is to be handled, but it is the big one.

And then take remaining commits out of https://github.com/OISF/suricata/pull/11321
And supply an example of an app-layer plugin

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
  

Pipeline 22322

File: src/output-json-netflow.c [hunk 1]
```diff
@@ -237,7 +239,9 @@ static void NetFlowLogEveToClient(JsonBuilder *js, Flow *f)
     jb_set_string(js, "start", timebuf1);
     jb_set_string(js, "end", timebuf2);
 
-    int32_t age = SCTIME_SECS(f->lastts) - SCTIME_SECS(f->startts);
+    DEBUG_VALIDATE_BUG_ON((int64_t)(SCTIME_SECS(f->lastts) - SCTIME_SECS(f->startts)) < INT32_MIN ||
+                          (int64_t)(SCTIME_SECS(f->lastts) - SCTIME_SECS(f->startts)) > INT32_MAX);
+    int32_t age = (int32_t)(SCTIME_SECS(f->lastts) - SCTIME_SECS(f->startts));
```
Comment (jufajardini) [inline src/output-json-netflow.c:244]: Since this is repeated a few times, couldn't it be wrapped in a function?

File: src/flow-private.h [hunk 2]
```diff
@@ -74,7 +74,7 @@ enum {
     FLOW_PROTO_MAX,
 };
 /* max used in app-layer (counters) */
-#define FLOW_PROTO_APPLAYER_MAX FLOW_PROTO_UDP + 1
+#define FLOW_PROTO_APPLAYER_MAX (FLOW_PROTO_UDP + 1)
```
Comment (jufajardini) [inline src/flow-private.h:77]: nit: shouldn't this one be part of https://github.com/OISF/suricata/pull/11701/commits/c32bc616ee5bd436c1a58c708b14bce841f12956 :P

**Review by jufajardini — COMMENTED** [review]
Didn't manage to review everything, leaving minor comments.


(the inversion protomap<->alproto for app-layer is mind fogging 😅  )

File: src/output-json-netflow.c (hunk 1 above)
Comment (catenacyber) [reply src/output-json-netflow.c:244]: Oops, this commit should not be in this PR

File: src/flow-private.h (hunk 2 above)
Comment (catenacyber) [reply src/flow-private.h:77]: Both are possible.
As a matter of facts, I did https://github.com/OISF/suricata/commit/c32bc616ee5bd436c1a58c708b14bce841f12956 after finding this parenthesis addition was needed for 897b6606e4ae116337ce9899d34c7ab9c6a14fda

**Comment by jufajardini** [comment]
> > (the inversion protomap<->alproto for app-layer is mind fogging 😅 )
> 
> Do you have a better C way to do this ?

Nothing that comes easily. Was just a silly comment.

**Comment by catenacyber** [comment]
> > > (the inversion protomap<->alproto for app-layer is mind fogging 😅 )
> > 
> > 
> > Do you have a better C way to do this ?
> 
> Nothing that comes easily. Was just a silly comment.

Not silly, I felt the pain, but found no other way...

**Pull request closed without merging**